modules/l4postgres/matcher.go

```diff
@@ -92,64 +52,133 @@ func (*MatchPostgres) CaddyModule() caddy.ModuleInfo {
 
 // Match returns true if the connection looks like the Postgres protocol.
 func (m *MatchPostgres) Match(cx *layer4.Connection) (bool, error) {
-	// Get bytes containing the message length
-	head := make([]byte, initMessageSizeLength)
-	if _, err := io.ReadFull(cx, head); err != nil {
-		return false, err
+	// Read message length (first 4 bytes)
+	lenBytes := make([]byte, lenFieldSize)
+	if _, err := io.ReadFull(cx, lenBytes); err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			return false, nil // Not enough data for PostgreSQL
+		}
+		return false, fmt.Errorf("reading message length: %w", err)
 	}
 
-	// Get actual message length
-	data := make([]byte, binary.BigEndian.Uint32(head)-initMessageSizeLength)
-	if _, err := io.ReadFull(cx, data); err != nil {
-		return false, err
+	// Parse and validate message length
+	msgLen := binary.BigEndian.Uint32(lenBytes)
+	if msgLen < minMessageLen {
+		return false, nil // Too small to be a valid PostgreSQL message
 	}
 
-	b := newMessageFromBytes(data)
+	// Calculate and validate payload length
+	payloadLen := msgLen - lenFieldSize
+	if payloadLen > maxPayloadSize {
+		return false, nil // Payload too large, reject to prevent DoS
+	}
 
-	// Check if it is a SSLRequest
-	code := b.ReadUint32()
-	if code == sslRequestCode {
-		return true, nil
+	// Read the payload
+	payload := make([]byte, payloadLen)
+	if _, err := io.ReadFull(cx, payload); err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			return false, nil // Incomplete message
+		}
+		return false, fmt.Errorf("reading payload: %w", err)
 	}
 
-	// Check supported protocol
-	if majorVersion := code >> 16; majorVersion < 3 {
-		return false, errors.New("pg protocol < 3.0 is not supported")
+	// Need at least 4 bytes for the code/version
```

Comment:
Hi! You do io.ReadFull above, i.e. len(payload) always equals payloadLen. Why don't you check payloadLen >= 4 before reading? I mean we could move this check up to optimize io reads.